# Pre-lecture

## 1.

In [2]:
import pandas as pd
url="https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df=pd.read_csv(url, encoding='latin1')
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

## 2.

### (1)

In [3]:
rows, columns=df.shape
print(f"Number of rows: {rows}, Number of columns: {columns}")

Number of rows: 391, Number of columns: 11


### (2)

**Definitions**:

*Observations*: 

Individual dataset entries which contain all kinds of information gathered for a single entry (rows). In the context of the villagers' dataset, it includes the information of the name, gender, species, birthday, etc of a single villager.

*Variables*: 

Kinds of information for all entries (columns). In the dataset, it would be all names, genders, species, etc of every villager.

## 3.

In [6]:
df.info

<bound method DataFrame.info of      row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye

In [10]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [11]:
df.describe(include='all')

row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391               391   
unique            92      388               391   
top     K.K. Country  wee one  villager-admiral   
freq              10        2                 1   
mean             NaN      NaN               NaN   
std              NaN      NaN               NaN   
min              NaN      NaN               NaN   
25%              NaN      NaN               NaN   
50%              NaN      NaN               NaN   
75%              NaN      NaN               NaN   
max              NaN      NaN               NaN   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [29]:
df.value_counts()

survived  pclass  sex     age   sibsp  parch  fare      embarked  class  who    adult_male  deck  embark_town  alive  alone
1         1       female  24.0  0      0      69.3000   C         First  woman  False       B     Cherbourg    yes    True     2
                          58.0  0      0      26.5500   S         First  woman  False       C     Southampton  yes    True     1
                          49.0  0      0      25.9292   S         First  woman  False       D     Southampton  yes    True     1
                                1      0      76.7292   C         First  woman  False       D     Cherbourg    yes    False    1
                          50.0  0      1      247.5208  C         First  woman  False       B     Cherbourg    yes    False    1
                                                                                                                              ..
                          16.0  0      0      86.5000   S         First  woman  False       B     Sout

## 4.

Using titanic.csv to asnwer this question

In [4]:
import pandas as pd
url="https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
titanic_df=pd.read_csv(url, encoding='latin1')

In [27]:
df.shape

(891, 15)

In [28]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [31]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [32]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### (a)

The discrepanccies between the sizes is different because ``df.shape`` is an attribute, and as such analyzes all columns, regardless of what kind of data they contain (float64, object, boolean, etc). ``df.describe()``, however, only analyzes float64 and int64. As ``df.describe`` ends with ``()``, it is thus a method and as such can only analyze data with numerical values.

### (b)

``df.shape`` gives the full size of the data (total number of rows), regardless of missing values. ``df.describe()``, however, only shows non-missing values, and as such may be lower than the value reported by ``df.shape``.

## 5.

``df.shape``, which does not end with ``()`` and is thus an attribute, analyzes all data regardless of their kind (float, object, int, etc) and does not account for missing values.

``df.describe()``, which does end with ``()`` and so is a method, analyzes only some kinds of data (float and int). It also takes missing values into consideration.

## ChatGPT Session Summary

"1. **Dataset Analysis**:
   - You initially provided a dataset URL, and I explained how to determine the number of rows and columns using the `df.shape` method in Python with `pandas`.
   - We discussed the definitions of **observations** (rows) and **variables** (columns) in the context of a dataset, where observations represent individual data entries, and variables represent the attributes being measured.
   
2. **Simple Summaries of Columns**:
   - I provided Python code to summarize a dataset using methods such as:
     - `df.info()` to get an overview of the dataset, including column data types and non-null counts.
     - `df.describe()` to get statistical summaries of numerical columns.
     - `df.value_counts()` to get the distribution of categorical variables.

3. **Switching to the Titanic Dataset**:
   - You switched to the Titanic dataset, and I provided updated code to summarize this new dataset, including explanations of key columns like "survived," "pclass," "age," etc.

4. **Discrepancies Between `df.shape` and `df.describe()`**:
   - We discussed potential discrepancies between the size of the dataset given by `df.shape` and what is reported by `df.describe()`. Specifically:
     - **Number of columns analyzed**: `df.describe()` analyzes only numeric columns by default, while `df.shape` reports all columns.
     - **Values in the "count" column**: `df.describe()` only counts non-missing values, so the count might be lower than the total number of rows.

Throughout the session, we focused on providing both explanations and Python code examples to help you work with and analyze datasets effectively."

## ChatGPT Session Log

Link: [ChatGPT STA130 - Week 1 - Homework - Pre-lecture Session](https://chatgpt.com/share/fb5f8339-18f3-406d-9c9d-0df69b738d71)

# Post-lecture

## 6.

**'count'**: 
Returns the total number of non-null entries in each column.

**'mean'**: 
Returns the arithmetric average of entries in a column.

**'std'**:
Returns the standard deviation of entries in a column.

**'min'**: 
Returns the minimum value entry value for each column.

**'25%'**: 
Returns the 25th percentile (first quartile (Q1)) of entries in a column. I.e., 25% of data in each column fall below this value.

**'50%'**: 
Returns the 50th percentile (median) of entries in a column. I.e., 50% of data in each column falls below this value and 50% are above this value.

**'75%'**: 
Returns the 75th percentile (third quartile (Q3)) of entries in a 
column. I.e., 75% of data in each column falls below this value.

**'max'**:
Returns the maximum value entry value for each column.

## 7.

In [62]:
import pandas as pd
url="https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df=pd.read_csv(url, encoding='latin1')

### (1)

``df.dropna()`` might be preferred over using ``del df['col']`` when, for example, when one wants to disregard observations that, because of missing data, may skew or disrupt data analysis. 

Using titanic.csv, an example of a "use case" in which using ``df.dropna()`` may be preffered is if one wanted to calculate the average age of all passengers. There are 177 missing age entries which interfere with finding the average age of all 890 passengers. This is because calculating the average age would not calculate the average age of 890 passengers, but rather the average age of 713 of those passengers. Using ``del df['col']`` would delete the entire age column as it has missing entries. However, the goal is to calculate the average age, and as such removing rows with missing entries will produce a more accurate picture.

In [54]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [55]:
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### (2)

Using ``del df['col']`` might be preferred over using ``df.dropna()`` if one wanted to include only columns with all information available to include all rows in a dataset, regardless of whether they have missing entries or not.

In titanic.csv, there are four columns with at least one missing entry. Rather than deleting rows with missing data, one could instead delete columns with missing data and thus still include all 890 passengers in their data analysis.

### (3)

Applying ``del df['col']`` before ``df.dropna()`` when both are used together could be important if one wanted to first delete some columns with missing values and then deal with the remaining missing entries using ``df.dropna()``.

### (4)

In [58]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [57]:
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [63]:
del df['age']
del df['deck']

In [66]:
df.dropna()

survived  pclass     sex  sibsp  parch     fare embarked   class    who  \
0           0       3    male      1      0   7.2500        S   Third    man   
1           1       1  female      1      0  71.2833        C   First  woman   
2           1       3  female      0      0   7.9250        S   Third  woman   
3           1       1  female      1      0  53.1000        S   First  woman   
4           0       3    male      0      0   8.0500        S   Third    man   
..        ...     ...     ...    ...    ...      ...      ...     ...    ...   
886         0       2    male      0      0  13.0000        S  Second    man   
887         1       1  female      0      0  30.0000        S   First  woman   
888         0       3  female      1      2  23.4500        S   Third  woman   
889         1       1    male      0      0  30.0000        C   First    man   
890         0       3    male      0      0   7.7500        Q   Third    man   

     adult_male  embark_town alive  alone  
0          True  Southampton    no  False  
1         False    Cherbourg   yes  False  
2         False  Southampton   yes   True  
3         False  Southampton   yes  False  
4          True  Southampton    no   True  
..          ...          ...   ...    ...  
886        True  Southampton    no   True  
887       False  Southampton   yes   True  
888       False  Southampton    no  False  
889        True    Cherbourg   yes   True  
890        True   Queenstown    no   True  

[889 rows x 13 columns]

I used ``del df['age']`` and ``del df['deck']`` first before using ``df.dropna()`` as, due to the high amount of missing data in the "age" and "deck" columns, I wanted to delete those two columns to impede them from skewing with any future analysis I may conduct. I then applied ``df.dropna()`` afterwards to remove the two rows with missing entries from the "embarked"/"embarked_town" column (same data used for both columns), as two entires are much less likely to skew with analyses while their respective columns still provide valuable data.

##  ChatGPT Session Summary

"1. **Column Definitions in Titanic Dataset**:
   - **`count`**: Number of non-null entries in each column.
   - **`mean`**: Arithmetic average of values.
   - **`std`**: Standard deviation of values.
   - **`25%`, `50%`, `75%`**: Percentiles of the data (25th, 50th, and 75th).
   - **`max`**: Maximum value in the data.

2. **`df.dropna()` vs. `del df['col']`**:
   - **`df.dropna()`**: Removes rows or columns with missing values (NaNs). It can be customized to drop rows or columns based on specific conditions.
   - **`del df['col']`**: Deletes a specific column from the DataFrame. It is used to remove the entire column, not based on missing values.

3. **Data Entry**:
   - Refers to a single piece of information or record within a data structure like lists, dictionaries, DataFrames, or databases.

4. **Use Case**:
   - In Python/statistical analysis, a use case refers to a specific scenario where Python tools are applied to solve a problem or perform tasks related to data analysis, such as data cleaning, exploratory analysis, hypothesis testing, and more.

5. **Dataset Specifics**:
   - **Titanic Dataset Columns**: 
     - `PassengerId`, `Survived`, `Pclass`, `Name`, `Sex`, `Age`, `SibSp`, `Parch`, `Ticket`, `Fare`, `Cabin`, `Embarked`.
   - **Difference Between Deck and Class**:
     - **Deck**: Refers to the physical level on the ship where a cabin is located.
     - **Class**: Refers to the level of service and accommodation quality (First, Second, Third).

6. **Handling Columns**:
   - **`del df['col']`**: Deletes a column entirely.
   - **`df.dropna()`**: Removes rows or columns with missing values based on specified conditions."

## ChatGPT Session Log

Link: [ChatGPT STA130 - Week 1 - Homework - Post-lecture Session](https://chatgpt.com/share/66e37550-46d8-8001-8f0b-34dcb86943b2)

## 8.

In [12]:
import pandas as pd
url="https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df=pd.read_csv(url, encoding='latin1')

### (1)

``df.groupby("col1")["col2"].describe()``

``df.groupby("col1")`` groups DataFrame (df) by the values of a specified column (``"col1"``)

``["col2"]`` selects another specified column (``"col2"``) from each group after grouping by ``"col1"``

``.describe()`` provides summary statistics defined in question 6.

In other words, ``df.groupby("col1")["col2"].describe()`` groups data by ``"col1"`` and, for each kind of statistics in that column, provide summary statistics based on ``"col2"``

In [17]:
# Example: Group columns "fare" by "class"

df.groupby("class")["age"].describe()

count       mean        std   min   25%   50%   75%   max
class                                                            
First   186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
Second  173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
Third   355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

Over here, passengers have been grouped by their class and have been, for each class, provided summary statistics based on their ``"age"`` column.

### (2)

In [7]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

``df.describe()`` has different values in the ``count`` for different data columns and as such capture something fundamentally different since it does not group columns together and rather deals with them seperately. On the other hand, ``df.groupby("col1")["col2"].describe()`` involves grouping and analyzing data from different columns.

### (3)

#### (A)

In [1]:
url="https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df=pd.read_csv(url, encoding='latin1')

NameError: name 'pd' is not defined

When I explained what I input and what the error code was, ChatGPT showed me in one simple step (with a built-in code saying, ``"import pandas as pd"``) what I needed to add to my input.

After I google the error code, I had to find the right "People also ask" drop-down search to see what I needed to add to my input. Even after I found it, it said what I had to do in English words and not ``"import pandas as pd"``.

#### (B)

In [9]:
import pandas as pd
url="titanics.csv"
df=pd.read_csv(url, encoding='latin1')

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

When I told ChatGPT my input and output, the first possible error it suggested to me was that the file name was not spelled correctly (``'titnaics.csv'`` instead of the correct name).

In Google, it took me a couple minutes to find a website that suggested the correct solution to this error.

#### (C)

In [6]:
DF.shape

NameError: name 'DF' is not defined

ChatGPT suggested at the beginning of its response that, because Python is case-sensitive, I was getting this error code because I typed ``df`` in all-caps.

Google first suggested that I check whether ``df`` was already defined. After a couple minutes of searching, I found the correct "case-sensitive" suggestion.

#### (D)

In [5]:
df.describe)

SyntaxError: unmatched ')' (2065609750.py, line 1)

ChatGPT immediately identified the missing parenthesis when I asked what was wrong with my code.

Google was also helpful with this error, as I was able to find the solution quickly.

#### (E)

In [9]:
df.isan().sum()

AttributeError: 'DataFrame' object has no attribute 'isan'

ChatGPT pointed out that I got this error because I mispelled ``'isna'`` or ``'isnull'``.

Google, however, only suggested that my function was mispelled.

#### (F)

In [10]:
df.groupby("Fare")["embark"].describe()

KeyError: 'Fare'

ChatGPT pointed out that one possible reason I got this error code was because I capitalized a letter that should not have been. In this case, the letter "F" should not have been capitalized, as the column ``'fare'`` does not have a capital "F".

Google also suggested that my error was in an unintentional capitalization. However, it did not explicitely say this but rather said that I got this error message because I input a key that does not exist.

#### (G)

In [18]:
df.groupby("fare")[embarked].describe()

NameError: name 'embarked' is not defined

ChatGPT said that I got this error message because the work "embarked" was not between quotes (single or double), and as such Python could not understand that it was supposed to reference a column.

Google suggested that my variable should be properly defined and assigned a variable. This is true to this case, but Google did not specifically mention that I got this error as a result of not including quotes.

----
Overall, I found ChatGPT to be much more helpful with finding errors in my code than Google. While both helped me find the mistakes, ChatGPT did so much quicker and pointed to the specific error(s) that were present, whereas Google referenced possible errors and general reasons as to why I was receiving these error messages.

### ChatGPT Session Summary

"1. **Loading the Titanic Dataset**: 
   - You initially loaded the Titanic dataset from a URL using `pandas`, but encountered some errors like `NameError` and `FileNotFoundError` when trying to reference the dataset or local files. We resolved these by ensuring `pandas` was imported correctly and by using the correct file paths and variable names.

2. **Shape and Summary Statistics**: 
   - We discussed how to use `df.shape` to check the dimensions of the DataFrame and how to use `df.describe()` to get descriptive statistics for the numerical columns.

3. **Groupby and Describe**:
   - You learned how `df.groupby('col1')['col2'].describe()` works by grouping data based on one column (`col1`) and providing descriptive statistics for another column (`col2`), with examples from the Titanic dataset.

4. **Common Errors and Fixes**:
   - We went over several common errors:
     - **Case Sensitivity**: Errors such as `KeyError: 'Fare'` were caused by incorrect capitalization. The correct column name was `fare`, not `Fare`.
     - **Accessing Columns**: You encountered `NameError` while accessing columns. The correct way to access a column is `df['column_name']` (with quotes).

5. **Checking Missing Values**:
   - We discussed how to check for missing values using `df.isna()` and how to interpret the output.

### Key Points:
- **Case Sensitivity**: Column names in `pandas` are case-sensitive.
- **Correct Syntax**: Always access DataFrame columns using `df['column_name']`.
- **Descriptive Statistics**: You can use `describe()` for summary statistics, and `groupby()` to break down statistics by categories."

### ChatGPT Session Log

Link: [ChatGPT STA130 - Week 1 - Homework - Post-lecture Session - Question 8](https://chatgpt.com/share/66e3b622-9740-8001-956c-34f3ec1e4ae5)

## 9.

Yes, I have interacted with the course textbook and other humans to help me understand the tutorial and lecture material.